### PR TITLE
chore: use the new process API in podman extension

### PR DESCRIPTION
**Pull request by vzhukovs**
### What does this PR do?

The following changes proposal provides a refactoring for an existing code to use the new process execution API.

### Screenshot/screencast of this PR

N/A

### What issues does this PR fix or reference?

need for #3174 

### How to test this PR?

Check that existing features related to podman communication works as before.

### CODE REVIEW

File: extensions/podman/src/macos-checks.ts [hunk 1]
```diff
@@ -67,23 +66,28 @@ export class MacPodmanInstallCheck extends BaseCheck {
...
+        return this.createSuccessfulResult();
+      }
     }
 
     // brew is installed, check if podman has been installed with brew
-    const runResult = await runCliCommand('brew', ['list', '--verbose', 'podman'], {
-      env: { HOMEBREW_NO_AUTO_UPDATE: '1', HOMEBREW_NO_ANALYTICS: '1' },
-    });
-
-    if (runResult.exitCode === 0) {
```
Comment (lstocchi) [inline extensions/podman/src/macos-checks.ts:81]: The previous check was if the exitCode was 0, now if it's different to 0.
I never used brew. I'm just highlighting it to be sure it was done on purpose.
Was this a bug? 
What i find awkward is the function name tat's returned. 
We check if podman was installed by brew, if it fails we return the `failure result` with the text "You have podman installed by brew" :thinking: But if it fails shouldn't it be "podman is not installed by brew"? ... 
BTW just a thought not related to this PR.
If you can confirm that the new check is correct, it's ok for me.

**Review by lstocchi — COMMENTED** [review]
LGTM by looking at the code, just one comment. I'll give it a quick test when i switch on windows.
Sonarcloud returns an error

File: extensions/podman/src/extension.ts [hunk 2]
```diff
@@ -79,7 +80,8 @@ export type MachineInfo = {
 
 async function updateMachines(provider: extensionApi.Provider): Promise<void> {
   // init machines available
-  const machineListOutput = await execPromise(getPodmanCli(), ['machine', 'list', '--format', 'json']);
+  const machineListOutput = (await extensionApi.process.exec(getPodmanCli(), ['machine', 'list', '--format', 'json']))
+    .stdout;
```
Comment (benoitf) [inline extensions/podman/src/extension.ts]: I think destructuring is easier to read
```suggestion
  const {stdout: machineListOutput} = await extensionApi.process.exec(getPodmanCli(), ['machine', 'list', '--format', 'json']);
```

File: extensions/podman/src/extension.ts [hunk 3]
```diff
@@ -138,21 +140,25 @@ async function updateMachines(provider: extensionApi.Provider): Promise<void> {
...
         if (isWindows()) {
-          socketPath = await execPromise(getPodmanCli(), [
-            'machine',
-            'inspect',
-            '--format',
-            '{{.ConnectionInfo.PodmanPipe.Path}}',
-            machineName,
-          ]);
+          socketPath = (
+            await extensionApi.process.exec(getPodmanCli(), [
```
Comment (benoitf) [inline extensions/podman/src/extension.ts]: probably need to use destructure as well {stdout: socketPath} = await ...

File: extensions/podman/src/extension.ts [hunk 4]
```diff
@@ -138,21 +140,25 @@ async function updateMachines(provider: extensionApi.Provider): Promise<void> {
...
+          ).stdout;
         } else {
-          socketPath = await execPromise(getPodmanCli(), [
-            'machine',
-            'inspect',
-            '--format',
-            '{{.ConnectionInfo.PodmanSocket.Path}}',
-            machineName,
-          ]);
+          socketPath = (
```
Comment (benoitf) [inline extensions/podman/src/extension.ts]: destructure

File: extensions/podman/src/extension.ts [hunk 5]
```diff
@@ -432,11 +438,13 @@ async function registerProviderFor(provider: extensionApi.Provider, machineInfo:
       await startMachine(provider, machineInfo, context, logger);
     },
     stop: async (context, logger): Promise<void> => {
-      await execPromise(getPodmanCli(), ['machine', 'stop', machineInfo.name], { context, logger });
+      await extensionApi.process.exec(getPodmanCli(), ['machine', 'stop', machineInfo.name], {
+        logger: new LoggerDelegator(context, logger),
```
Comment (benoitf) [inline extensions/podman/src/extension.ts:517]: Could you explain why we need a 'LoggerDelegator'
I would expect that we won't need 3rd party utilities

File: extensions/podman/src/extension.ts [hunk 6]
```diff
@@ -838,7 +855,8 @@ export async function activate(extensionContext: extensionApi.ExtensionContext):
 // Restart current machine
 export async function restartCurrentMachine(): Promise<void> {
   // Find the machines
-  const machineListOutput = await execPromise(getPodmanCli(), ['machine', 'list', '--format', 'json']);
+  const machineListOutput = (await extensionApi.process.exec(getPodmanCli(), ['machine', 'list', '--format', 'json']))
+    .stdout;
```
Comment (benoitf) [inline extensions/podman/src/extension.ts]: destructure

File: extensions/podman/src/extension.ts [hunk 7]
```diff
@@ -853,15 +871,16 @@ export async function restartCurrentMachine(): Promise<void> {
...
 }
 
 // Function that checks to see if the default machine is running and return a boolean
 export async function findRunningMachine(): Promise<string> {
   let runningMachine: string;
 
   // Find the machines
-  const machineListOutput = await execPromise(getPodmanCli(), ['machine', 'list', '--format', 'json']);
+  const machineListOutput = (await extensionApi.process.exec(getPodmanCli(), ['machine', 'list', '--format', 'json']))
+    .stdout;
```
Comment (benoitf) [inline extensions/podman/src/extension.ts]: destructure

Also as it's like the 5 times that I see this command being called, maybe it needs its own function like

```
podmanMachineList(): string {
...
}
```

and we don't copy/paste `extensionApi.process.exec(getPodmanCli(), ['machine', 'list', '--format', 'json']` 5times in a row

File: extensions/podman/src/macos-checks.ts [hunk 8]
```diff
@@ -67,23 +66,28 @@ export class MacPodmanInstallCheck extends BaseCheck {
...
+        env: { HOMEBREW_NO_AUTO_UPDATE: '1', HOMEBREW_NO_ANALYTICS: '1' },
+      });
+    } catch (err) {
+      if ((err as extensionApi.RunError) && err.exitCode !== 0) {
+        return this.createFailureResult(
+          'You have podman installed with "brew", run "brew update && brew upgrade podman" to install new version',
+          'Brew Documentation',
+          'https://docs.brew.sh/Manpage#upgrade-options-outdated_formulaoutdated_cask-',
+        );
+      }
```
Comment (benoitf) [inline extensions/podman/src/macos-checks.ts]: agree with @lstocchi 

the error here is if we see that podman is installed using brew that we need to report an error
(so basically if no error then create an error)

File: extensions/podman/src/podman-cli.ts [hunk 9]
```diff
@@ -56,87 +54,13 @@ export function getCustomBinaryPath(): string | undefined {
...
-}
-
 export interface InstalledPodman {
   version: string;
 }
 
 export async function getPodmanInstallation(): Promise<InstalledPodman | undefined> {
   try {
-    const versionOut = await execPromise(getPodmanCli(), ['--version']);
+    const versionOut = (await process.exec(getPodmanCli(), ['--version'])).stdout;
```
Comment (benoitf) [inline extensions/podman/src/podman-cli.ts]: wondering if we don't need some small commands like getPodmanCliVersion() , getPodmanMachineList(), etc

File: extensions/podman/src/podman-install.ts [hunk 10]
```diff
@@ -305,9 +306,13 @@ class MacOSInstaller extends BaseInstaller {
       const pkgPath = path.resolve(getAssetsFolder(), `podman-installer-macos-${pkgArch}-v${podmanTool.version}.pkg`);
       try {
         if (fs.existsSync(pkgPath)) {
-          const runResult = await runCliCommand('open', [pkgPath, '-W']);
-          if (runResult.exitCode !== 0) {
+          try {
+            await extensionApi.process.exec('open', [pkgPath, '-W']);
+          } catch (err) {
             throw new Error(runResult.stdErr);
+            if ((err as extensionApi.RunError) && err.exitCode) {
```
Comment (benoitf) [inline extensions/podman/src/podman-install.ts]: Don't know why we need to check for the exitCode as if we're inside the catch block it means exit Code was != 0

File: extensions/podman/src/podman-install.ts [hunk 11]
```diff
@@ -409,9 +414,11 @@ class VirtualMachinePlatformCheck extends BaseCheck {
   async execute(): Promise<extensionApi.CheckResult> {
     try {
       // set CurrentUICulture to force output in english
-      const res = await execPromise('powershell.exe', [
-        '(Get-WmiObject -Query "Select * from Win32_OptionalFeature where InstallState = \'1\'").Name | select-string VirtualMachinePlatform',
-      ]);
+      const res = (
```
Comment (benoitf) [inline extensions/podman/src/podman-install.ts]: destructure

File: extensions/podman/src/podman-install.ts [hunk 12]
```diff
@@ -473,15 +480,18 @@ class WSL2Check extends BaseCheck {
     return str;
   }
   private async isUserAdmin(): Promise<boolean> {
-    const res = await execPromise('powershell.exe', [
-      '$null -ne (whoami /groups /fo csv | ConvertFrom-Csv | Where-Object {$_.SID -eq "S-1-5-32-544"})',
-    ]);
+    const res = (
```
Comment (benoitf) [inline extensions/podman/src/podman-install.ts]: destructure

File: extensions/podman/src/util.ts [hunk 13]
```diff
@@ -67,44 +60,34 @@ export function getAssetsFolder(): string {
   }
 }
 
-export function runCliCommand(command: string, args: string[], options?: RunOptions): Promise<SpawnResult> {
-  return new Promise((resolve, reject) => {
-    let output = '';
-    let err = '';
-    let env = Object.assign({}, process.env); // clone original env object
+export class LoggerDelegator implements Logger {
```
Comment (benoitf) [inline extensions/podman/src/util.ts:79]: it's unclear why we would need this function.

All should be 'easy' for consumers of the API.

File: extensions/podman/src/macos-checks.ts [hunk 14]
```diff
@@ -67,23 +66,28 @@ export class MacPodmanInstallCheck extends BaseCheck {
...
   async execute(): Promise<extensionApi.CheckResult> {
     // we need to check if brew is installed to avoid unexpected error
-    const brewResult = await runCliCommand('which', ['brew']);
-    if (brewResult.exitCode !== 0) {
-      // brew is not installed so do not check if podman has been installed with brew
-      return this.createSuccessfulResult();
+    try {
+      await extensionApi.process.exec('which', ['brew']);
+    } catch (err) {
+      if ((err as extensionApi.RunError) && err.exitCode !== 0) {
```
Comment (benoitf) [inline extensions/podman/src/macos-checks.ts]: why do we check the exit code ? if it's catch block it means we have not exitCode being 0

**Review by benoitf — CHANGES_REQUESTED** [review]
some 'small' remarks

but for brew check it has changed the logic

File: extensions/podman/src/macos-checks.ts (hunk 1 above)
Comment (vzhukovs) [reply extensions/podman/src/macos-checks.ts:81]: Here was a problem with correct processing exit codes, now, the code has been changed the way how it was originally designed.

File: extensions/podman/src/extension.ts (hunk 2 above)
Comment (vzhukovs) [reply extensions/podman/src/extension.ts]: updated

File: extensions/podman/src/extension.ts (hunk 3 above)
Comment (vzhukovs) [reply extensions/podman/src/extension.ts]: updated

File: extensions/podman/src/extension.ts (hunk 4 above)
Comment (vzhukovs) [reply extensions/podman/src/extension.ts]: updated

File: extensions/podman/src/extension.ts (hunk 5 above)
Comment (vzhukovs) [reply extensions/podman/src/extension.ts:517]: added description to the LoggerDelegator class: https://github.com/containers/podman-desktop/pull/3315/files#diff-62f1e4c4e882bb44ccd1b237c25717b7bab687b6e4f2fbe41576294a555209f4R63

File: extensions/podman/src/extension.ts (hunk 6 above)
Comment (vzhukovs) [reply extensions/podman/src/extension.ts]: updated

File: extensions/podman/src/extension.ts (hunk 7 above)
Comment (vzhukovs) [reply extensions/podman/src/extension.ts]: extracted call to the dedicate function: `getJSONMachineList`

File: extensions/podman/src/macos-checks.ts (hunk 8 above)
Comment (vzhukovs) [reply extensions/podman/src/macos-checks.ts]: reviewed and updated the statements, agreed, there was a problem. a bit reworked, now the logic should not be changed with the original one

File: extensions/podman/src/podman-cli.ts (hunk 9 above)
Comment (vzhukovs) [reply extensions/podman/src/podman-cli.ts]: as far as this method is used ones, not sure, that we need a dedicate function for that, unless this call will be used somewhere else

File: extensions/podman/src/podman-install.ts (hunk 10 above)
Comment (vzhukovs) [reply extensions/podman/src/podman-install.ts]: updated and simplified a bit

File: extensions/podman/src/podman-install.ts (hunk 11 above)
Comment (vzhukovs) [reply extensions/podman/src/podman-install.ts]: updated

File: extensions/podman/src/podman-install.ts (hunk 12 above)
Comment (vzhukovs) [reply extensions/podman/src/podman-install.ts]: updated

File: extensions/podman/src/util.ts (hunk 13 above)
Comment (vzhukovs) [reply extensions/podman/src/util.ts:79]: added a js doc section with the description

File: extensions/podman/src/macos-checks.ts (hunk 14 above)
Comment (vzhukovs) [reply extensions/podman/src/macos-checks.ts]: updated

**Review by lstocchi — CHANGES_REQUESTED** [review]
On Windows i face this error when i try to start/stop a podman machine 

image

**Comment by vzhukovs** [comment]
@lstocchi I've fixed the bug with this error, re-checked on windows, there is no such error anymore

**Review by benoitf — APPROVED** [review]
checked by stopping/starting podman etc and it worked

**Review by cdrage — COMMENTED** [review]
Testing this out and extensions are able to load just as before / there are no issues or errors, etc.

Will need a rebase as a few files have changed, but LGTM!

**Review by cdrage — APPROVED** [review]

**Review by lstocchi — APPROVED** [review]
Sorry @vzhukovs i was on PTO. 
Tested again, it works fine. Thanks!!

**Comment by cdrage** [comment]
e2e smoke tasks failing are unrelated, but after a rebase / re-run of the test, we should be good @vzhukovs

**Comment by vzhukovs** [comment]
> e2e smoke tasks failing are unrelated, but after a rebase / re-run of the test, we should be good @vzhukovs

checking that

**Pull request merged**